pkg/network/apis/network/validation/validation.go

```diff
@@ -50,29 +50,49 @@ func ValidateClusterNetwork(clusterNet *networkapi.ClusterNetwork) field.ErrorLi
-		//In the case that a user manually makes a clusterNetwork object with clusterNet.Network and clusterNet.HostubnetLength at least make sure they are valid values
-		clusterIPNet, err := validateCIDRv4(clusterNet.Network)
-		if err != nil {
-			allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, err.Error()))
+	serviceIPNet, err := validateCIDRv4(clusterNet.ServiceNetwork)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, err.Error()))
+	}
+
+	if len(clusterNet.ClusterNetworks) == 0 {
+		// legacy ClusterNetwork; old fields must be set
+		if clusterNet.Network == "" {
+			allErrs = append(allErrs, field.Required(field.NewPath("network"), "network must be set (if clusterNetworks is empty)"))
+		} else if clusterNet.HostSubnetLength == 0 {
+			allErrs = append(allErrs, field.Required(field.NewPath("hostsubnetlength"), "hostsubnetlength must be set (if clusterNetworks is empty)"))
 		} else {
+			clusterIPNet, err := validateCIDRv4(clusterNet.Network)
+			if err != nil {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, err.Error()))
+			}
 			maskLen, addrLen := clusterIPNet.Mask.Size()
 			if clusterNet.HostSubnetLength > uint32(addrLen-maskLen) {
-				allErrs = append(allErrs, field.Invalid(field.NewPath("hostsubnetlength"), clusterNet.HostSubnetLength, "subnet length is too large for clusterNetwork"))
+				allErrs = append(allErrs, field.Invalid(field.NewPath("hostsubnetlength"), clusterNet.HostSubnetLength, "subnet length is too large for cidr"))
 			} else if clusterNet.HostSubnetLength < 2 {
 				allErrs = append(allErrs, field.Invalid(field.NewPath("hostsubnetlength"), clusterNet.HostSubnetLength, "subnet length must be at least 2"))
 			}
+
+			if (clusterIPNet != nil) && (serviceIPNet != nil) && configapi.CIDRsOverlap(clusterIPNet.String(), serviceIPNet.String()) {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("serviceNetwork"), clusterNet.ServiceNetwork, "service network overlaps with cluster network"))
+			}
+		}
+	} else {
+		// "new" ClusterNetwork
+		if clusterNet.Name == networkapi.ClusterNetworkDefault {
+			if clusterNet.Network != clusterNet.ClusterNetworks[0].CIDR {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("network"), clusterNet.Network, "network must be identical to clusterNetworks[0].cidr"))
+			}
+			if clusterNet.HostSubnetLength != clusterNet.ClusterNetworks[0].HostSubnetLength {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("hostsubnetlength"), clusterNet.HostSubnetLength, "hostsubnetlength must be identical to clusterNetworks[0].hostSubnetLength"))
+			}
+		} else if clusterNet.Network != "" || clusterNet.HostSubnetLength != 0 {
+			if clusterNet.Network != clusterNet.ClusterNetworks[0].CIDR || clusterNet.HostSubnetLength != clusterNet.ClusterNetworks[0].HostSubnetLength {
```

Comment:
Do we need to enforce this condition for non default cluster network? May be we shouldn't care.

Reply:
What would it mean to have Network set but not equal to ClusterNetworks[0]?

(I had suggested a long time back in the original PR that we should keep using Network/HostSubnetLength for the primary network, and only use ClusterNetworks for *additional* networks. But it's too late to change that now.)

Reply:
Okay, OpenShift sdn only cares about default cluster network but based on what we described in swagger doc for Network/ClusterNetworks, now it seems reasonable to expect network to match clusterNetworks[0] for other use cases as part of api validation.